### DISCUSSION TITLE
Is there a way to use any markdown-it plugins, like Shiki?

### DISCUSSION THREAD

**Discussion by aretrosen** [markdoc/markdoc#517] [Q&A]
Today, I was writing some plugins for `markdoc`, similar to the ones I used to use for `remark` and `rehype`. One of the plugins I used to use was `@shikijs/rehype` for Shiki's highlighting. I noticed an example of code highlighting using Prism, but since Prism appears to be unmaintained, I tried to use Shiki's `markdown-it` plugin instead, as I saw that `markdoc` is based on `markdown-it`. However, I couldn't make it work. This is what I tried:
```ts
const tokenizer = new Markdoc.Tokenizer();
// TODO: Can I just hook up plugins this way?
// for (const plugin of config.plugins) {
// 	tokenizer['parser'].use(plugin);
// }
// test with shiki
tokenizer["parser"].use(
  await Shiki({
    themes: {
      light: "vitesse-light",
      dark: "vitesse-dark",
    },
  })
);
```

If using Shiki this way isn't possible, I wrote a `Hast` to `Tag` solution, since Shiki can output `HAST`. It works, but I'm unsure if it's comprehensive. Has anyone used `remark` or `rehype` in the past and can confirm whether the `Hast` to `Tag` conversion is correct and/or comprehensive?

```ts
export function hastToMarkdocTag(
  root: Root
): RenderableTreeNode | RenderableTreeNode[] {
  const stack: [Root | Element | Text, RenderableTreeNode[]][] = [[root, []]];
  let result: RenderableTreeNode | RenderableTreeNode[] = [];

  while (stack.length > 0) {
    const [node, parentChildren] = stack.pop()!;

    if (node.type === "element") {
      const children: RenderableTreeNode[] = [];
      const tag = new Markdoc.Tag(node.tagName, node.properties, children);
      parentChildren.push(tag);

      for (let i = node.children.length - 1; i >= 0; i--) {
        stack.push([node.children[i] as Root | Element | Text, children]);
      }
    } else if (node.type === "text") {
      parentChildren.push(node.value);
    } else if (node.type === "root") {
      for (let i = node.children.length - 1; i >= 0; i--) {
        stack.push([
          node.children[i] as Root | Element | Text,
          result as RenderableTreeNode[],
        ]);
      }
    }
  }

  return root.type === "root" && result.length === 1 ? result[0] : result;
}
```
Also, let me know if there are any simpler solutions I might have overlooked, similar to the one available for `prismjs`?

**Comment by segphault** [comment]
We generally don't encourage this, but you can subclass the tokenizer and call `this.parser.use` in the constructor in order to add markdown-it plugins. Some plugins may not work as expected if they produce tokens that aren't recognized by Markdoc's parser. Programmatically generating a Markdoc AST from the output of your syntax highlighter is probably a more idiomatic approach.
